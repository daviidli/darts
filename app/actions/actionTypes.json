{
	"SET_SERIAL_PORT": "SET_SERIAL_PORT",
	"SET_WAIT_TIME": "SET_WAIT_TIME",
	"SET_PLAYERS": "SET_PLAYERS",
	"ADD_PLAYER": "ADD_PLAYER",
	"REMOVE_PLAYER": "REMOVE_PLAYER",
	"CHANGE_PLAYER": "CHANGE_PLAYER",
	"SET_MAX_DARTS": "SET_MAX_DARTS",
	"SET_MAX_ROUNDS": "SET_MAX_ROUNDS",
	"SET_CURRENT_PLAYER": "SET_CURRENT_PLAYER",
	"SET_CURRENT_DART": "SET_CURRENT_DART",
	"SET_CURRENT_ROUND": "SET_CURRENT_ROUND",
	"INITIALIZE_ROUNDS": "INITIALIZE_ROUNDS",
	"SET_ROUND": "SET_ROUND",
	"SET_TOTALS": "SET_TOTALS",
	"SET_WINNER": "SET_WINNER",
	"SET_WAITING": "SET_WAITING"
}
