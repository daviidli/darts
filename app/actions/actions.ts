import actions from './actions.json';

export const setSerialPort = (port: string) => ({
	type: actions.SET_SERIAL_PORT,
	port
});

export const test = () => {
	return {};
};
