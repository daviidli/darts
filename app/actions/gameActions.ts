import actions from './actionTypes.json';

export const setPlayers = (players: string[]) => ({
	type: actions.SET_PLAYERS,
	players
});

export const setPlayer = (player: string, index: number) => ({
	type: actions.SET_PLAYER,
	player,
	index
});

export const addPlayer = (player: string) => ({
	type: actions.ADD_PLAYER,
	player
});

export const removePlayer = (player: string) => ({
	type: actions.REMOVE_PLAYER,
	player
});

export const setCurrentPlayer = (index: number) => ({
	type: actions.SET_CURRENT_PLAYER,
	index
});

export const setMaxRounds = (maxRounds: number) => ({
	type: actions.SET_MAX_ROUNDS,
	maxRounds
});

// todo: set rounds

export const addRound = (round: string[]) => ({
	type: actions.ADD_ROUND,
	round
});

export const setMaxThrows = (maxThrows: number) => ({
	type: actions.SET_MAX_THROWS,
	maxThrows
});

export const setCurrentRound = (currentRound: number) => ({
	type: actions.SET_CURRENT_ROUND,
	currentRound
});

export const setCurrentThrow = (currentThrow: string[]) => ({
	type: actions.SET_CURRENT_THROW,
	currentThrow
});

export const setTotals = (totals: number[]) => ({
	type: actions.SET_TOTALS,
	totals
});
