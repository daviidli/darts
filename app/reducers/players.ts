import { AnyAction } from 'redux';
import actions from '../actions/actionTypes.json';

const players = (state = [], action: AnyAction) => {
	switch (action.type) {
		case actions.SET_PLAYERS:
			return action.players;
		default:
			return state;
	}
};

export default players;
