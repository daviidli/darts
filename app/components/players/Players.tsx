import React from 'react';
import {
	List,
	ListItem,
	TextField,
	ListItemSecondaryAction,
	IconButton,
	Button
} from '@material-ui/core';
import { connect } from 'react-redux';
import { MdDelete } from 'react-icons/md';
import { stateType, playerType } from '../../reducers/types';
import {
	addPlayer as addPlayerAction,
	removePlayer as removePlayerAction,
	changePlayer as changePlayerAction
} from '../../actions/gameActions';

type Props = {
	players: playerType[];
	addPlayer: (player: string) => {};
	removePlayer: (index: number) => {};
	changePlayer: (players: string, index: number) => {};
};

const Players = (props: Props) => {
	const { players, addPlayer, removePlayer, changePlayer } = props;
	return (
		<>
			<List>
				{players.map((player, i) => (
					<ListItem key={`player-${player.id}`}>
						<TextField
							label={`Player ${i + 1}`}
							value={player.name}
							onChange={e => changePlayer(e.target.value, i)}
							color="secondary"
						/>
						<ListItemSecondaryAction>
							<IconButton
								edge="end"
								onClick={() => removePlayer(i)}
							>
								<MdDelete />
							</IconButton>
						</ListItemSecondaryAction>
					</ListItem>
				))}
			</List>
			<Button
				variant="contained"
				color="secondary"
				onClick={() => addPlayer(`player ${players.length + 1}`)}
			>
				Add player
			</Button>
		</>
	);
};

const mapStateToProps = (state: stateType) => ({ players: state.players });
const mapDispatchToProps = {
	addPlayer: addPlayerAction,
	removePlayer: removePlayerAction,
	changePlayer: changePlayerAction
};

export default connect<any, any, any, stateType>(
	mapStateToProps,
	mapDispatchToProps
)(Players);
