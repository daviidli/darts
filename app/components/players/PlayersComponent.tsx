import React from 'react';
import {
	List,
	ListItem,
	TextField,
	ListItemSecondaryAction,
	IconButton,
	Button
} from '@material-ui/core';
import { MdDelete } from 'react-icons/md';
import { playerType } from '../../reducers/types';

export type Props = {
	players: playerType[];
	addPlayer: (player: string) => {};
	removePlayer: (index: number) => {};
	changePlayer: (players: string, index: number) => {};
};

const PlayersComponent = (props: Props) => {
	const { players, addPlayer, removePlayer, changePlayer } = props;
	return (
		<>
			<List>
				{players.map((player, i) => (
					<ListItem key={`player-${player.id}`}>
						<TextField
							label={`Player ${i + 1}`}
							value={player.name}
							onChange={e => changePlayer(e.target.value, i)}
							color="secondary"
						/>
						<ListItemSecondaryAction>
							<IconButton
								edge="end"
								onClick={() => removePlayer(i)}
							>
								<MdDelete />
							</IconButton>
						</ListItemSecondaryAction>
					</ListItem>
				))}
			</List>
			<Button
				variant="contained"
				color="secondary"
				onClick={() => addPlayer(`player ${players.length + 1}`)}
			>
				Add Player
			</Button>
		</>
	);
};

export default PlayersComponent;
