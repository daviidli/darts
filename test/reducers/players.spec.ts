import { AnyAction } from 'redux';
import actions from '../../app/actions/actionTypes.json';
import players from '../../app/reducers/players';

describe('players reducer', () => {
	it('should handle initial state', () => {
		const action = { type: 'someType' };
		expect(players(undefined, action)).toEqual([]);
	});

	it('should return same state', () => {
		const state = [{ id: '1', name: 'player1' }];
		const action: AnyAction = {
			type: 'someType'
		};
		expect(players(state, action)).toEqual(state);
	});

	it('should set players', () => {
		const state = [];
		const newPlayers = [{ id: '1', name: 'player1' }];
		const action: AnyAction = {
			type: actions.SET_PLAYERS,
			players: newPlayers
		};
		expect(players(state, action)).toEqual(newPlayers);
	});

	it('should add a player', () => {
		const state = [];
		const action: AnyAction = {
			type: actions.ADD_PLAYER,
			player: 'unicorn'
		};
		const result = players(state, action);
		expect(result).toHaveLength(1);
		expect(result[0]).toMatchSnapshot({
			id: expect.any(String), // jest was unable to mock uuid/v4
			name: 'unicorn'
		});
	});

	it('should remove a player', () => {
		const state = [{ id: '1', name: 'player1' }];
		const action: AnyAction = {
			type: actions.REMOVE_PLAYER,
			index: 0
		};
		expect(players(state, action)).toEqual([]);
	});

	it('should try to remove a player out of range', () => {
		const state = [{ id: '1', name: 'player1' }];
		const action: AnyAction = {
			type: actions.REMOVE_PLAYER,
			index: 10
		};
		expect(players(state, action)).toEqual(state);
	});

	it("should change a player's name", () => {
		const state = [{ id: '1', name: 'player1' }];
		const action: AnyAction = {
			type: actions.CHANGE_PLAYER,
			player: 'unicorn',
			index: 0
		};
		expect(players(state, action)).toEqual([{ id: '1', name: 'unicorn' }]);
	});

	it("should change a player's name out of range", () => {
		const state = [{ id: '1', name: 'player1' }];
		const action: AnyAction = {
			type: actions.CHANGE_PLAYER,
			player: 'unicorn',
			index: 10
		};
		expect(players(state, action)).toEqual(state);
	});
});
