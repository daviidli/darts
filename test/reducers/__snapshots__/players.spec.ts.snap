// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`players reducer should add a player 1`] = `
Object {
  "id": Any<String>,
  "name": "unicorn",
}
`;
